File: lbr_examples/scripts/lbr_command_node.py
#!/usr/bin/python3
import math
from copy import deepcopy

import rclpy
from rclpy import qos
from rclpy.node import Node

from lbr_fri_msgs.msg import LBRCommand, LBRState


class LBRSinusoidalNode(Node):
    amplitude_: float
    omega_: float
    t0_: float
    initial_lbr_state_: LBRState

    def __init__(self, node_name: str = "lbr_sinusoidal_node"):
        super().__init__(node_name)

        self.declare_parameter("amplitude", math.pi / 4.0)
        self.declare_parameter("period", 20.0)

        self.amplitude_ = float(self.get_parameter("amplitude").value)
        self.omega_ = 2*math.pi / float(self.get_parameter("period").value)
        self.t0_ = None
        self.initial_lbr_state_ = None

        self.lbr_state_sub_ = self.create_subscription(
            LBRState,
            "/lbr_state",
            self.lbr_state_sub_cb_,
            qos.qos_profile_system_default,
        )
        self.lbr_command_pub_ = self.create_publisher(
            LBRCommand, "/lbr_command", qos.qos_profile_system_default
        )
        self.lbr_command_timer_ = self.create_timer(0.01, self.timer_cb_)

    def lbr_state_sub_cb_(self, msg: LBRState) -> None:
        if not self.initial_lbr_state_:
            self.t0_ = float(self.get_clock().now().nanoseconds)
            self.initial_lbr_state_ = msg

    def timer_cb_(self):
        if not self.initial_lbr_state_:
            return
        command = LBRCommand()
        command.joint_position = deepcopy(self.initial_lbr_state_.measured_joint_position)
        t = (float(self.get_clock().now().nanoseconds) - self.t0_) / 1.0e9
        command.joint_position[3] -= self.amplitude_ * math.sin(self.omega_ * t)
        command.joint_position[4] -= self.amplitude_ * math.sin(self.omega_ * t)
        command.joint_position[5] += self.amplitude_ * math.sin(self.omega_ * t)
        command.joint_position[6] += self.amplitude_ * math.sin(self.omega_ * t)
        self.lbr_command_pub_.publish(command)


def main(args=None):
    rclpy.init(args=args)

    lbr_sinusoidal_node = LBRSinusoidalNode()
    rclpy.spin(lbr_sinusoidal_node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: lbr_demos/src/lbr_command_node.cpp
#include <chrono>
#include <cmath>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"

#include "lbr_fri_msgs/msg/lbr_command.hpp"
#include "lbr_fri_msgs/msg/lbr_state.hpp"

class LBRCommandNode : public rclcpp::Node {
public:
  LBRCommandNode(const std::string &node_name = "lbr_command_node")
      : rclcpp::Node(node_name), initial_lbr_state_(nullptr) {
    declare_parameter("amplitude", M_PI_4);
    declare_parameter("period", 20.);

    amplitude_ = get_parameter("amplitude").as_double();
    omega_ = 2. * M_PI / get_parameter("period").as_double();

    lbr_state_sub_ = create_subscription<lbr_fri_msgs::msg::LBRState>(
        "/lbr_state", rclcpp::SensorDataQoS(),
        std::bind(&LBRCommandNode::lbr_state_sub_cb_, this, std::placeholders::_1));

    lbr_command_pub_ =
        create_publisher<lbr_fri_msgs::msg::LBRCommand>("/lbr_command", rclcpp::SensorDataQoS());

    lbr_command_timer_ = create_wall_timer(std::chrono::milliseconds(10),
                                           std::bind(&LBRCommandNode::timer_cb_, this));
  };

protected:
  void lbr_state_sub_cb_(const lbr_fri_msgs::msg::LBRState::SharedPtr lbr_state) {
    if (!initial_lbr_state_) {
      t0_ = static_cast<double>(now().nanoseconds());
      initial_lbr_state_ = lbr_state;
    }
  }

  void timer_cb_() {
    if (!initial_lbr_state_) {
      return;
    }
    lbr_fri_msgs::msg::LBRCommand lbr_command;
    lbr_command.joint_position = initial_lbr_state_->measured_joint_position;
    double t = static_cast<double>(now().nanoseconds() - t0_) / 1.e9;
    lbr_command.joint_position[6] += amplitude_ * std::sin(omega_ * t);
    lbr_command_pub_->publish(lbr_command);
  }

  double amplitude_;
  double omega_;
  double t0_;

  lbr_fri_msgs::msg::LBRState::SharedPtr initial_lbr_state_;

  rclcpp::TimerBase::SharedPtr lbr_command_timer_;
  rclcpp::Subscription<lbr_fri_msgs::msg::LBRState>::SharedPtr lbr_state_sub_;
  rclcpp::Publisher<lbr_fri_msgs::msg::LBRCommand>::SharedPtr lbr_command_pub_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<LBRCommandNode>());
  rclcpp::shutdown();
  return 0;
}
